### PR TITLE
Demand estimation

### PR DESCRIPTION
**Pull request by imcurtis**
## Background
I took the feedback from @kmagida110 and @bamader and made a separate folder with different files for field lookups as well as a script for the demand estimations. So now, there is both a script and a Jupyter notebook.

## GitHub Issue
Replace this text with the GitHub issue link.
**NOTE: Please don't mark the issue as closed by this pull request!  Issues are to be closed manually within Zenhub upon merge & prod deploy.**

## Associated PRs
https://github.com/ctoec/data-visualization/tree/town-demographics

## Validation Plan
I changed the lookup files to contain only the variables I needed for my demand estimation. But I did keep Brandon's script the pulls data from the census. I know the next step will likely involve me merging my demand estimates with what Kyle is doing.

## Additional Context
If there's anything additional information that those reviewing or validating this work should be aware of, please specify it here.

### CODE REVIEW

File: data_sets/expanded_town_demographics/combination_field_lookups.txt [hunk 1]
```diff
@@ -22,4 +22,7 @@ rent paid for non-owned housing units; B25063; 27; 27; no rent paid; 3,4,5,6,7,8
 rent as percentage of income; B25070; 10; 10; 50+%; 2,3 + 4,5,6 + 7,8,9; <15%, 15-29%, 30-49%
 home value of owned/mortgaged housing units; B25075; 27; 21,22,23,24; 300k - 400k, 400k - 500k, 500k - 750k, 750k - 1m; 2,3,4,5,6,7,8,9 + 10,11,12,13,14 + 15,16,17,18 + 19,20 + 25,26,27; < 50k, 50k - 100k, 100k - 200k, 200k - 300k, 1m+
 monthly housing costs; B25104; 17; 17,16; no housing cost, 3k+; 2,3,4,5,6 + 7,8,9,10,11 + 12,13 + 14,15; < 500, 500-1000, 1k - 2k, 2k - 3k
-computers and internet; B28003; 6; 6,5; no computer, computer but no internet; 3,4; computer with internet
\ No newline at end of file
+computers and internet; B28003; 6; 6,5; no computer, computer but no internet; 3,4; computer with internet
+median income by family size;B19119; 7; 1; median income by household; 2,3,4 + 5,6,7; 2-4 family size, 4+ family size
```
Comment (kmagida110) [inline data_sets/expanded_town_demographics/combination_field_lookups.txt:26]: I'll leave it up to @bamader but did we want these in a separate file?

File: src/data_integration/demand_estimation/demand_estimate_script.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import censusdata
+import pandas as pd
+import numpy as np
+
+OUT_FILE = "town_demandEstimated_data.csv"
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/demand_estimate_script.py:5]: Nit, add an underscore and lowercase the E

File: src/data_integration/demand_estimation/demand_estimate_script.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+'''
+Define a function for each of the four demand estimates already written in the Jupyter 
+notebook.
+'''
+
+# read in data
+
+town = pd.read_csv('town_demand_data.csv')
+
+def demand_estimate_1(female, male):
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/demand_estimate_script.py:16]: Can you have more descriptive names than 1-4?

File: src/data_integration/demand_estimation/demand_estimate_script.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+	return ((female + male)*0.5).apply(np.ceil)
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/demand_estimate_script.py:18]: What are female and male here? Pandas Series? Add docs

File: src/data_integration/demand_estimation/demand_estimate_script.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+# read in data
+
+town = pd.read_csv('town_demand_data.csv')
+
+def demand_estimate_1(female, male):
+
+	return ((female + male)*0.5).apply(np.ceil) 
+	#town.to_csv(OUT_FILE, index=False)
+
+town['demand_estimate'] = demand_estimate_1(town['pop_female_under_5'], town['pop_male_under_5'])
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/demand_estimate_script.py:21]: Separate code from functions, anything that is actually running in the script should be below all the functions

File: src/data_integration/demand_estimation/demand_estimate_script.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+def ratio_poverty (level_1, level_2, level_3, total_pop):
+
+	return (level_1 + level_2 + level_3)/total_pop
+
+town['ratio_poverty'] = ratio_poverty(town['ratio of income to poverty level - < 0.5'],  
+town['ratio of income to poverty level - 0.51 - 1.0'], town['ratio of income to poverty level - 1.01 - 2.0'],
+town['total population'])
+
+def demand_estimate_2(ratio, female, male):
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/demand_estimate_script.py:31]: You can just have a ratio of pop function that multiplies by a ratio and for demand estimate 1 just pass .5 as a ratio.

File: src/data_integration/demand_estimation/demand_estimate_script.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+	return (ratio * (female + male)).apply(np.ceil)
+	
+town['demand_estimate_2'] = demand_estimate_2(town['ratio_poverty'], town['pop_female_under_5'], town['pop_male_under_5'])
+
+def lowerIncome(housing, total_pop):
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/demand_estimate_script.py:37]: use snake case here instead lower_income

File: src/data_integration/demand_estimation/demand_estimate_script.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	
+town['children_below_200%_poverty'] = below_200_poverty(town['children under 6 ratio of income to poverty level - < 0.5'], 
+town['children under 6 ratio of income to poverty level - 0.51 - 1.0'],
+town['children under 6 ratio of income to poverty level - 1.01 - 2.0'])
+
+def all_parents_working(both, dad_only, mom_only, all):
+
+	return (both + dad_only + mom_only)/all
+
+town['%all_parents_working'] = (town['children under 6 by parent employment status - both parents - in labor force'] +
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/demand_estimate_script.py:63]: Nit, spell out percent

File: src/data_integration/demand_estimation/field_demand_lookup.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
+import censusdata as cd
```
Comment (kmagida110) [inline src/data_integration/demand_estimation/field_demand_lookup.py:1]: If we are moving this to a new folder, and I think we should, then we should delete the old files

File: src/data_integration/unmet_needs/census_data/.idea/.gitignore [hunk 10]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
```
Comment (kmagida110) [inline src/data_integration/unmet_needs/census_data/.idea/.gitignore:1]: Remove this file, it should be ignored by gitignore anyways

Comment (kmagida110) [reply src/data_integration/unmet_needs/census_data/.idea/.gitignore:1]: ie. add this .idea/ folder to gitignore again

**Review by kmagida110 — CHANGES_REQUESTED** [review]
Some small nits but the models look good. Also, on the notebook, let's not print the tables to cut down on size, the head should be fine to get a sense of what is there

File: data_sets/expanded_town_demographics/combination_field_lookups.txt (hunk 1 above)
Comment (bamader) [reply data_sets/expanded_town_demographics/combination_field_lookups.txt:26]: Yes I think we should definitely keep these separate. Sorry for being unclear on that Alex. My intention was to have a directory as follows:

```
data_sets/
    expanded_town_demographics/
        single_field_lookups
        combination_lookups
    demand_estimation/
        single_field_lookups
        combination_field_lookups
```

So we want each data set that we create to have its own two lookup files. Does that make sense?

File: src/data_integration/demand_estimation/demand_estimate_script.py (hunk 5 above)
Comment (bamader) [reply src/data_integration/demand_estimation/demand_estimate_script.py:21]: +10000 to this ^^^ . I find it's a very small styling thing that personally improves readability for me by orders of magnitude. (I am also guilty of doing this in scripts that aren't finished, so I'll have to go back and fix this too before I push anything up)

File: src/data_integration/demand_estimation/demand_estimate_script.py (hunk 7 above)
Comment (bamader) [reply src/data_integration/demand_estimation/demand_estimate_script.py:37]: +1

**Comment by bamader** [comment]
Alex, I second most of Kyle's organizational points. The big things to me are:
1. we need to preserve all the lookup files for each data set we make; this keeps our data set directories standardized and allows exact replication in the future if we need to; so, you should move your lookup fields to new textfiles in another sub-directory of data_sources
2. we want the output csv to be as clean and simple as possible. If you know in advance that we're going to be using model 4 (either because it's the best, most relevant, or one we care most about), I think the code's output should be a CSV with just two columns: town, and that model's results. This could definitely be a personal choice, and if you'd rather keep all models, that's fine too. But what I really think is necessary is some way of indicating which of the models is most important and the one that should be used as the source of truth.

File: data_sets/expanded_town_demographics/combination_field_lookups.txt (hunk 1 above)
Comment (imcurtis) [reply data_sets/expanded_town_demographics/combination_field_lookups.txt:26]: That makes sense, so should I create a new PR where I have a folder within the data_sets repo?

**Pull request closed without merging**